### DISCUSSION TITLE
What is IAlternateEqualityComparer?

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#104423] [Q&A] [answered]
I've started seeing this interface pop up in PRs but I'm not sure why it exists or what it is used for.
Is there any documentation I can reference?

**Comment by En3Tho** [comment, marked as answer]
@TonyValenti Basically an interface in addition to IEqualityComparer that says "hey, you've got some data of type X there and I can compare it with data of type Y.

With it you can look up values in string based Dictionaries using spans for example. I believe this was one of the main motivations for it.

**Reply by stephentoub** [reply]
Exactly. The built-in string comparers all implement `IAlternateEqualityComparer<ReadOnlySpan<char>, string>`, for example.

**Reply by TonyValenti** [reply]
Ahh!  So it is basically IEqualityComparer<TLeft, TRight>
?

**Reply by stephentoub** [reply]
Yes
